### PR TITLE
Stop retry flow for error handling (log trigger)

### PR DESCRIPTION
**Pull request by amirylm**
AUTO-9005

#### Description

The idea is to stop retry ceremony after 6 rounds (first 5 retries 1 second, 1 retry 5 second) and then invoke error handler.

**NOTES** 
- does not address conditional triggers that should fallback after 1 round
- In this PR we will return `timeout` once the retry round is 6, but in followup PRs we will make a call to err handler through `executeCallback()`

#### Changes

- added error codes
- return error code from `DoRequest`
- return timeout upon high round retries count
- added placeholder for next PRs to prepare args for `checkErrHandler`

#### Testing

- Tested retryInterval (with timeout once reached)
- Tested error codes
- Followup PRs should add integration tests

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v03/request.go [hunk 1]
```diff
@@ -61,9 +61,9 @@ func NewClient(mercuryConfig mercury.MercuryConfigProvider, httpClient mercury.H
 	}
 }
 
-func (c *client) DoRequest(ctx context.Context, streamsLookup *mercury.StreamsLookup, pluginRetryKey string) (encoding.PipelineExecutionState, encoding.UpkeepFailureReason, [][]byte, bool, time.Duration, error) {
+func (c *client) DoRequest(ctx context.Context, streamsLookup *mercury.StreamsLookup, pluginRetryKey string) (encoding.PipelineExecutionState, encoding.UpkeepFailureReason, [][]byte, bool, time.Duration, encoding.ErrCode, error) {
 	if len(streamsLookup.Feeds) == 0 {
-		return encoding.NoPipelineError, encoding.UpkeepFailureReasonInvalidRevertDataInput, [][]byte{}, false, 0 * time.Second, fmt.Errorf("invalid revert data input: feed param key %s, time param key %s, feeds %s", streamsLookup.FeedParamKey, streamsLookup.TimeParamKey, streamsLookup.Feeds)
+		return encoding.NoPipelineError, encoding.UpkeepFailureReasonInvalidRevertDataInput, [][]byte{}, false, 0 * time.Second, 0, fmt.Errorf("invalid revert data input: feed param key %s, time param key %s, feeds %s", streamsLookup.FeedParamKey, streamsLookup.TimeParamKey, streamsLookup.Feeds)
```
Comment (shileiwill) [inline core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v03/request.go:66]: nit: encoding.ErrCodeNil?

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v03/request.go [hunk 2]
```diff
@@ -154,26 +157,31 @@ func (c *client) multiFeedsRequest(ctx context.Context, ch chan<- mercury.Mercur
 			case http.StatusUnauthorized:
 				retryable = false
 				state = encoding.UpkeepNotAuthorized
+				errCode = encoding.HttpToErrCode(resp.StatusCode)
```
Comment (shileiwill) [inline core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v03/request.go:160]: seems we can save some lines by moving this assignment outside of switch-case

**Review by shileiwill — APPROVED** [review]
LGTM: added some nitpicking comments.

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v02/request.go [hunk 3]
```diff
@@ -148,6 +151,7 @@ func (c *client) singleFeedRequest(ctx context.Context, ch chan<- mercury.Mercur
 				c.lggr.Warnf("at block %s upkeep %s received status code %d for feed %s", sl.Time.String(), sl.UpkeepId.String(), httpResponse.StatusCode, sl.Feeds[index])
 				retryable = true
 				state = encoding.MercuryFlakyFailure
+				errCode = encoding.HttpToErrCode(httpResponse.StatusCode)
```
[the remark this answers is not in the document]
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v02/request.go:154]: We want the error code just in specific scenarios, so just to be sure I've preferred to do it specifically when needed.

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v03/request.go (hunk 2 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v03/request.go:160]: We want the error code just in specific scenarios, being explicit here seems more safe.

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/streams/streams.go [hunk 4]
```diff
@@ -237,20 +238,34 @@ func (s *streams) CheckCallback(ctx context.Context, values [][]byte, lookup *me
...
 	if err != nil {
 		checkResults[i].Retryable = retryable
 		checkResults[i].RetryInterval = retryInterval
 		checkResults[i].PipelineExecutionState = uint8(state)
 		checkResults[i].IneligibilityReason = uint8(reason)
+		retryTimeout := retryInterval == mercury.RetryIntervalTimeout
+		if retryTimeout {
+			// in case of retry timeout, setting retryable to false
+			checkResults[i].Retryable = false
+		}
```
[the remark this answers is not in the document]
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/streams/streams.go:259]: Will be tested in followup PRs

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/mercury.go [hunk 5]
```diff
@@ -86,7 +87,7 @@ type HttpClient interface {
 }
 
 type MercuryClient interface {
-	DoRequest(ctx context.Context, streamsLookup *StreamsLookup, pluginRetryKey string) (encoding.PipelineExecutionState, encoding.UpkeepFailureReason, [][]byte, bool, time.Duration, error)
+	DoRequest(ctx context.Context, streamsLookup *StreamsLookup, pluginRetryKey string) (encoding.PipelineExecutionState, encoding.UpkeepFailureReason, [][]byte, bool, time.Duration, encoding.ErrCode, error)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/mercury.go:90]: Can we add a comment here on what all these fields are supposed to represent. 

It is confusing that we have multiple similar fields such as pipeline execution state, failure reason, error code and there are constraints that some of these things can't happen together. (e.g. if error code is returned then state should be no error etc)

Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/mercury.go:90]: If retry interval returned here is `RetryIntervalTimeout`, then would we always expect a non zero error code?

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v02/request.go [hunk 6]
```diff
@@ -53,11 +53,11 @@ func NewClient(mercuryConfig mercury.MercuryConfigProvider, httpClient mercury.H
 	}
 }
 
-func (c *client) DoRequest(ctx context.Context, streamsLookup *mercury.StreamsLookup, pluginRetryKey string) (encoding.PipelineExecutionState, encoding.UpkeepFailureReason, [][]byte, bool, time.Duration, error) {
+func (c *client) DoRequest(ctx context.Context, streamsLookup *mercury.StreamsLookup, pluginRetryKey string) (encoding.PipelineExecutionState, encoding.UpkeepFailureReason, [][]byte, bool, time.Duration, encoding.ErrCode, error) {
 	resultLen := len(streamsLookup.Feeds)
 	ch := make(chan mercury.MercuryData, resultLen)
 	if len(streamsLookup.Feeds) == 0 {
-		return encoding.NoPipelineError, encoding.UpkeepFailureReasonInvalidRevertDataInput, [][]byte{}, false, 0 * time.Second, fmt.Errorf("invalid revert data input: feed param key %s, time param key %s, feeds %s", streamsLookup.FeedParamKey, streamsLookup.TimeParamKey, streamsLookup.Feeds)
+		return encoding.NoPipelineError, encoding.UpkeepFailureReasonInvalidRevertDataInput, [][]byte{}, false, 0 * time.Second, encoding.ErrCodeNil, fmt.Errorf("invalid revert data input: feed param key %s, time param key %s, feeds %s", streamsLookup.FeedParamKey, streamsLookup.TimeParamKey, streamsLookup.Feeds)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v02/request.go:60]: technically this is also a user error and we can surface this in callback.

However, I'm not sure if we want to go through all 0.2 cases, properly define it and also test all cases. So one option is to just always return 0 error code here for 0.2 to satisfy the interface, while properly handle it for 0.3

cc @anirudhwarrier

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/streams/streams.go [hunk 7]
```diff
@@ -237,20 +238,34 @@ func (s *streams) CheckCallback(ctx context.Context, values [][]byte, lookup *me
 }
 
 func (s *streams) DoMercuryRequest(ctx context.Context, lookup *mercury.StreamsLookup, checkResults []ocr2keepers.CheckResult, i int) ([][]byte, error) {
-	state, reason, values, retryable, retryInterval, err := encoding.NoPipelineError, encoding.UpkeepFailureReasonInvalidRevertDataInput, [][]byte{}, false, 0*time.Second, fmt.Errorf("invalid revert data input: feed param key %s, time param key %s, feeds %s", lookup.FeedParamKey, lookup.TimeParamKey, lookup.Feeds)
+	state, reason, values, retryable, retryInterval, errCode, err := encoding.NoPipelineError, encoding.UpkeepFailureReasonInvalidRevertDataInput, [][]byte{}, false, 0*time.Second, encoding.ErrCodeNil, fmt.Errorf("invalid revert data input: feed param key %s, time param key %s, feeds %s", lookup.FeedParamKey, lookup.TimeParamKey, lookup.Feeds)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/streams/streams.go:241]: are these initial values used anywhere? I see we always overwrite them with either 0.2.DoRequest or 0.3.DoRequest, maybe simple to just declare them as vars?

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/mercury.go (hunk 5 above)
Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/mercury.go:90]: e.g. what would be the difference between error code and error

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v03/request.go [hunk 8]
```diff
@@ -154,26 +157,31 @@ func (c *client) multiFeedsRequest(ctx context.Context, ch chan<- mercury.Mercur
...
 				retryable = true
 				state = encoding.MercuryFlakyFailure
+				errCode = encoding.HttpToErrCode(resp.StatusCode)
 				return fmt.Errorf("%d", http.StatusPartialContent)
 			case http.StatusOK:
 				// continue
 			default:
 				retryable = false
 				state = encoding.InvalidMercuryRequest
+				errCode = encoding.ErrCodeBadRequest
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v03/request.go:184]: as of current implementation, retryInterval will need to be -1 for this error code to be handled

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v02/request.go (hunk 6 above)
Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v02/request.go:60]: for e.g. there are some errors like 0.2 permissioning which are done upstream and not in 0.2 module

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/streams/streams.go (hunk 7 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/streams/streams.go:241]: we could have declare some of the vars, but some has initial value and thats why they are defined like this

File: core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v02/request.go (hunk 6 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/mercury/v02/request.go:60]: We can decide in followup PRs if we want exclude v0.2

**Pull request merged**